libs/pilight/protocols/433.92/fanju.c

```diff
@@ -0,0 +1,283 @@
+		if(fanju->raw[i] > AVG_PULSE) {
+			binary[x++] = 1;
+		} else {
+			binary[x++] = 0;
+		}
+	}
+
+	// 4x SYNC '0' + 1x HEAD '1'
+	header = binToDecRev(binary, 0, 4);
+	if(header != 1) {
+		logprintf(LOG_ERR, "fanju: parsecode - invalid header %d", header);
+		return;
+	}
+
+	id = binToDecRev(binary, OFFSET, OFFSET + 7);
+	checksum = binToDecRev(binary, OFFSET + 8, OFFSET + 11);
+	battery = binary[13];
+	temp_fahrenheit = (double) binToDecRevUl(binary, OFFSET + 16, OFFSET + 27);
+	temp_celsius = ((temp_fahrenheit - 0x4C4) * 5) / 9;
+	temperature = temp_celsius / 10;
+	humidity_10 = binToDecRev(binary, OFFSET + 28, OFFSET + 31);
+	humidity = (double) binToDecRev(binary, OFFSET + 32, OFFSET + 35);
+	humidity += humidity_10 * 10;
+	channel = binToDecRev(binary, OFFSET + 38, OFFSET + 39);
+
+	// move channel to the checksum position
+	x = 0;
+	for(i=OFFSET + 0; i < OFFSET + 8; i++) {
+		binary_cpy[x++] = binary[i];
+	}
+	for(i=OFFSET + 36; i < OFFSET + 40; i++) {
+		binary_cpy[x++] = binary[i];
+	}
+	for(i=OFFSET + 12; i < OFFSET + 36; i++) {
+		binary_cpy[x++] = binary[i];
+	}
+	// verify checksum
+	mask = 0xC;
+	checksum_calc = 0x0;
+	for(i=0; i < 36; i++) {
+		bit = mask & 0x1;
+		mask >>= 1;
+		if(bit == 0x1) {
+			mask ^= 0x9;
+		}
+		if(binary_cpy[i] == 1) {
+			checksum_calc ^= mask;
+		}
+	}
+	if(checksum != checksum_calc) {
+		logprintf(LOG_ERR, "fanju: parsecode - invalid checksum: %d  calc: %d", checksum, checksum_calc);
+		return;
+    }
+
+	struct settings_t *tmp = settings;
+	while(tmp) {
+		if(fabs(tmp->id-id) < EPSILON) {
+			humi_offset = tmp->humi;
+			temp_offset = tmp->temp;
+			break;
+		}
+		tmp = tmp->next;
+	}
+	temperature += temp_offset;
+	humidity += humi_offset;
+
+	if(channel != 4) {
+        fanju->message = json_mkobject();
+        json_append_member(fanju->message, "id", json_mknumber(id, 0));
+        json_append_member(fanju->message, "temperature", json_mknumber(temperature, 1));
+        json_append_member(fanju->message, "humidity", json_mknumber(humidity, 1));
+        json_append_member(fanju->message, "battery", json_mknumber(battery, 0));
+        json_append_member(fanju->message, "channel", json_mknumber(channel, 0));
+	}
+}
+
+static int checkValues(struct JsonNode *jvalues) {
+	struct JsonNode *jid = NULL;
+
+	if((jid = json_find_member(jvalues, "id"))) {
+		struct settings_t *snode = NULL;
+		struct JsonNode *jchild = NULL;
+		struct JsonNode *jchild1 = NULL;
+		double id = -1;
+		int match = 0;
+
+		jchild = json_first_child(jid);
+		while(jchild) {
+			jchild1 = json_first_child(jchild);
+			while(jchild1) {
+				if(strcmp(jchild1->key, "id") == 0) {
+					id = jchild1->number_;
+				}
+				jchild1 = jchild1->next;
+			}
+			jchild = jchild->next;
+		}
+
+		struct settings_t *tmp = settings;
+		while(tmp) {
+			if(fabs(tmp->id - id) < EPSILON) {
+				match = 1;
+				break;
+			}
+			tmp = tmp->next;
+		}
+
+		if(match == 0) {
+			if((snode = MALLOC(sizeof(struct settings_t))) == NULL) {
```

Comment:
Alignment

Reply:
I'm so sorry, has been fixed now (and I should not be mixing Vim and TextWrangler)